[style] Introduce rules for polyglot headers

mcy opened the pull request:
Fixes https://github.com/lowRISC/opentitan/issues/1934

Still missing text about banning statics-in-inline; @lenary should I just stick that in the inline section like you suggested? It's technically a C-specific section. Also, please let me know if there's any other restrictions around inline we should be adding.

gkelly commented on doc/rm/c_cpp_coding_style.md:
````diff
@@ -139,6 +139,31 @@ Example:
+}  // extern "C"
+#endif
+```
+Moreover, all `#includes` must also be of other polyglot headers.
````
Doesn't this force considerations on other header writers after the fact? If you're including a header that wasn't intended to be polyglot (which is allowed from the third line of this change) then you don't have to write `extern` decls.

I think it might be easier to just make a requirement that headers for C libraries must carry `extern C` declarations.

mcy replied:
Sorry, I think I wasn't precise in my wording. Here's the intended meaning:
- Headers *may* be polyglot.
- A polyglot header *must* only include other polyglot headers.
- A polyglot header's body, after the includes, must be wrapped in `extern "C"`.

Exactly what did I write that is inconsistent with that? What can I add to clarify?

mcy commented on doc/rm/c_cpp_coding_style.md:
```diff
@@ -139,6 +139,31 @@ Example:
+***Headers intended to be included from both languages must contain `extern` guards; `#include`s should not be wrapped in `extern "C" {}`.***
+
+A *polyglot header* is a header file that can be safely included in either a `.c` or `.cc` file.
+In particular, this means that the file does not depend on any of the places where C and C++ semantics disagree.
```
Done.

mcy commented on doc/rm/c_cpp_coding_style.md:
```diff
@@ -139,6 +139,31 @@ Example:
+extern "C" {
+#endif
+
+// Polyglot declarations...
```
Hmm, yeah. I definitely don't think we should be conditioning anything on `__cplusplus` other than the `extern` bit. I'll just go with "Declarations...".

lenary commented on doc/rm/c_cpp_coding_style.md:
````diff
@@ -139,6 +139,31 @@ Example:
+}  // extern "C"
+#endif
+```
+Moreover, all `#includes` must also be of other polyglot headers.
````
Note to match our discussion: system headers are not required to be polyglot.

mcy replied:
Actually, I was thinking about this... at a minimum, we probably want `stdbool.h` to become an empty file `#ifdef __cplusplus`. I don't think that affects what you said, but does that seem right to you? Right now I think that including `stdbool.h` silently breaks some implicit conversions around `bool` in C++.

gkelly commented:
Do we have any headers that we won't want to make polyglot headers?

lenary commented:
> Do we have any headers that we won't want to make polyglot headers?

My preference would be "as many as possible", because I think this C/C++ mix does make things more complex.

Concretely, I think we've got quite a few libraries and the like that are only ever expected to run on-device (most of the `sw/device` tree excluding `sw/device/lib/{dif,base,testing}`), and I'm not sure we need to add the complexity of making those polyglot headers.

gkelly commented on doc/rm/c_cpp_coding_style.md:
````diff
@@ -139,6 +139,32 @@ Example:
+}  // extern "C"
+#endif
+```
+Moreover, all non-system `#includes` must also be of other polyglot headers.
+(In other words, all C system headers may be assumed to be polyglot, even if they lack guards.)
````
The language here suggests that we won't ever write a non-polyglot header, doesn't it?

lenary replied:
I think the intention here is `...all non-system includes in polyglot headers must…`.

mcy replied:
Done.

mcy commented:
@lenary This should go in a followup, but we're going to need to add some kind of header, like `stdstatic_assert.h` or something, to provide a C/C++-independent macro for `static_assert`, seeing as C names it `_Static_assert`, while in C++ it's `static_assert`. I.e., something like
```
#ifdef __cplusplus
#define static_assert static_assert
#else
#define static_assert _Static_assert
#endif
```
Thoughts?

lenary commented:
Or we shouldn't have static_assert in headers?

lenary commented:
I cannot find where we discussed it, but this PR needs to also update the guidance about inline functions to prevent the use of `static` local variables in inline functions, as this is known to be somewhere where C and C++ oppose each other in meaning.

mcy commented:
I mean, we could, but having `_Static_assert` for ensuring that the sizes of structs are correct (I ran into this when refactoring bootstrap.c). But, sure, if you think it's going to be an issue we can ban it in polyglot headers.

mcy commented:
Nevermind; I forgot that `assert.h`, in C and only C, is responsible for providing `static_assert`. Oops. =)

mcy commented:
@lenary the text is now updated to reflect that. Incidentally, `static` was all-round banned from headers already, but I underscored it nonetheless. I also added an example list of things that are different across C and C++, to try to get the point across that this is subtle and and something to be avoided.

I also added a carveout for vendored in stuff, which is not going to follow this rule but should otherwise be extern C'd.

lenary commented:
I had a question about autogenerated headers for @mcy, and I think we've agreed that autogenerated headers should have the `#ifdef __cplusplus` polyglot wrappers, as the `*_regs.h` are used in C++ tests. the `top_earlgrey.h` file is used from C++, but that might change (it probably shouldn't be used in DIF mock tests), so I'm less sure about that - but it probably makes sense to err on the side of caution and include it.
